File: py/20_ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
            input_chars = list(s)
            valid_pairs = {
                '[': ']',
                '{': '}',
                '(': ')'
            }

            # First character must be an opening bracket
            if not input_chars or input_chars[0] not in valid_pairs:
                return False

            # Must have an even number of characters
            if len(input_chars) % 2 != 0:
                return False

            # Stack to track expected closing brackets
            expected_stack = [valid_pairs[input_chars[0]]]

            for char in input_chars[1:]:
                if char in valid_pairs:
                    expected_stack.append(valid_pairs[char])
                elif expected_stack and char == expected_stack[-1]:
                    expected_stack.pop()
                else:
                    return False

            return len(expected_stack) == 0

File: cpp/20-ValidParentheses.cpp
#include <string>
#include <stack>
#include <unordered_map>
class Solution {
public:
    bool isValid(std::string s) {
        if (s.empty()) return false;

        std::unordered_map<char, char> valid_pairs = {
            {'[', ']'},
            {'{', '}'},
            {'(', ')'}
        };

        // First character must be an opening bracket
        if (valid_pairs.find(s[0]) == valid_pairs.end()) {
            return false;
        }

        // Must have an even number of characters
        if (s.length() % 2 != 0) {
            return false;
        }

        std::stack<char> expected_stack;
        expected_stack.push(valid_pairs[s[0]]);

        for (size_t i = 1; i < s.length(); ++i) {
            char c = s[i];
            if (valid_pairs.find(c) != valid_pairs.end()) {
                expected_stack.push(valid_pairs[c]);
            } else if (!expected_stack.empty() && c == expected_stack.top()) {
                expected_stack.pop();
            } else {
                return false;
            }
        }

        return expected_stack.empty();
    }
};